include/wil/win32_helpers.h

```diff
@@ -531,6 +531,105 @@ namespace wil
             FreeLibraryAndExitThread(thisModule, 0);
         });
     }
+
+    struct module_reference_for_new_thread
+    {
+        bool free_on_exit = false;
+        HMODULE module_ref = nullptr;
+
+        module_reference_for_new_thread()
+        {
+            FAIL_FAST_IF(!::GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, L"", &module_ref));
+        }
+
+        ~module_reference_for_new_thread()
+        {
+            reset();
+        }
+
+        module_reference_for_new_thread(module_reference_for_new_thread&& other)
+        {
+            wistd::exchange(free_on_exit, other.free_on_exit);

```

Comment:
Unless I'm going crazy, you want these to be `swap`. `exchange` just assigns the new value without setting/clearing the assigned-from value. I.e. `other` will still have its destructor run and you'll have a double free of `module_ref`

Reply:
Or perhaps you meant `free_on_exit = wistd::exchange(other.free_on_exit, false)` etc?